### DISCUSSION TITLE
Being able to ignore all unmapped source or target properties

### DISCUSSION THREAD

**Discussion by PaulVrugt** [riok/mapperly#677] [Q&A]
We are migrating from Automapper to mapperly. And we want to get errors about missing source or target member mappings, to makes sure that all members that need to be mapped are mapped. So we enabled the severity:

```
[*.cs]
dotnet_diagnostic.RMG012.severity = error # Unmapped target member
dotnet_diagnostic.RMG020.severity = error # Unmapped source member
```

this works, and we can real time errors if something is wrong. However, most of the time we want either all source or all target properties to be mapped (in automapper we could determine this by using the `MemberList.Source` or `MemberList.Target`). If we would have an attribute we could set to simply set the ignore ALL unmapped source/target properties, we wouldn't have to add a million `MapperIgnoreTarget` or `MapperIgnoreSource` attributes to mapping functions. Something like:

```c#
[MapperIgnoreAllMissingTargetMappings]
public partial Target MapToTarget(Source source)
```

Edit:

I came up with a workaround that is not ideal but it works, so I'm using it for now. We still have set the severity to error for the unmapped source and target members, but we have created 2 mappers, and use the `SuppressMessage` attribute on them to suppress either the unmapped source or unmapped target members:

```
[System.Diagnostics.CodeAnalysis.SuppressMessage("Mapper", "RMG012:Target member is not mapped to any source member", Justification = "When mapping from dto to entity, only the source members need to be mapped")]
```

**Comment by PaulVrugt** [comment]
I just found out the workaround doesn't actually work, since the compiler ignores my `SuppressMessage` attributes

**Comment by PaulVrugt** [comment]
So I would be willing to contribute to add this feature. I made some POC code in mapperly that supports this, but is nowhere near production ready. But I would like to have some indication on of you are willing to allow such an addition to the library or not, before putting more effort into this and showing the poc implementation to get a sense of the chosen solution it what you want it to be.

And you don't want to have this kind of functionality in the library, I would like to know what the alternatives are to this, because without this, we are not able to use mapperly in our project.

Why not? You might ask:
We want to force all members mapping to be checked using the RMG012 and RMG020 diagnostics to avoid unexpected mapping issues when someone renames a property at either end of the mapping classes. However, some mapping targets (or sources) have A LOT of properties that are not to be mapped, and we don't have to add a ignore attribute for each of those properties (and future properties) because this would imply a lot of overhead.

**Reply by TimothyMakkison** [reply]
For what it's worth I think this sounds like a reasonable addition. I'd be happy to add this if it gets approval from one of the riok guys ie @latonz or @CommonGuy.

I don't think the feature will require a massive change. I'd add it by adding something like `[MapperIgnoreAllSource]` and then iterating through each member adding each one to the ignore list. `[MapProperty]` will map ignored members.

**Reply by PaulVrugt** [reply]
cool. I took a different approach in my poc, by simply not emitting the diagnostic when respectively the `ignoreallsource` or `mapperignorealltarget` attributes are set, but either way would work. Hopefully the riok guys agree :)

**Reply by latonz** [reply]
Thank you for opening this discussion. I created an issue with a proposed API: https://github.com/riok/mapperly/issues/687. Feel free to add your feedback to the proposed API. Would be very happy to accept a PR implementing this 😊 If you start working on it, please ping me and I'll assign the issue to you.

**Comment by latonz** [comment]
This is available as a preview release `v3.3.0-next.1`.
